modules/apps/headless/headless-form/headless-form-impl/src/main/java/com/liferay/headless/form/internal/helper/UploadFileHelper.java

```diff
@@ -0,0 +1,205 @@
+			formField -> Objects.equals(formField.getType(), "document_library")
+		).map(
+			field -> {
+				Stream<DDMFormFieldValue> ddmFormFieldValuesStream =
+					ddmFormFieldValues.stream();
+
+				return ddmFormFieldValuesStream.filter(
+					value -> Objects.equals(field.getName(), value.getName())
+				).collect(
+					Collectors.toList()
+				);
+			}
+		).forEach(
+			values -> {
+				try {
+					for (DDMFormFieldValue ddmFormFieldValue : values) {
+						FileEntry fileEntry = _dlAppService.getFileEntry(
+							_getFileEntryId(ddmFormFieldValue));
+
+						JSONObject jsonObject =
+							JSONFactoryUtil.createJSONObject();
+
+						jsonObject.put(
+							"fileEntryId", fileEntry.getFileEntryId()
+						).put(
+							"groupId", fileEntry.getGroupId()
+						).put(
+							"title", fileEntry.getTitle()
+						).put(
+							"type", fileEntry.getMimeType()
+						).put(
+							"uuid", fileEntry.getUuid()
+						).put(
+							"version", fileEntry.getVersion()
+						);
+
+						DDMFormField ddmFormField =
+							ddmFormFieldValue.getDDMFormField();
+
+						String jsonValue = jsonObject.toString();
+
+						Value value = new UnlocalizedValue(jsonValue);
+
+						if (ddmFormField.isLocalizable()) {
+							value = new LocalizedValue();
+
+							value.addString(
+								value.getDefaultLocale(), jsonValue);
+						}
+
+						ddmFormFieldValue.setValue(value);
+					}
+				}
+				catch (Exception e) {
+					_log.error(e.getMessage());
+
+					throw new BadRequestException(e);
+				}
+			}
+		);
+	}
+
+	public FormDocument toDocument(FileEntry fileEntry) throws Exception {
+		FileVersion fileVersion = fileEntry.getFileVersion();
+
+		return new FormDocument() {
+			{
+				contentUrl = _dlurlHelper.getPreviewURL(
+					fileEntry, fileVersion, null, "");
+				encodingFormat = fileEntry.getMimeType();
+				fileEntryId = fileEntry.getFileEntryId();
+				fileExtension = fileEntry.getExtension();
+				siteId = fileEntry.getGroupId();
+				sizeInBytes = fileEntry.getSize();
+				title = fileEntry.getTitle();
+			}
+		};
+	}
+
+	public FileEntry uploadFile(
+			DDMFormInstance ddmFormInstance, MultipartBody multipartBody)
+		throws IOException, PortalException {
+
+		MediaForm mediaForm = multipartBody.getValueAsInstance(
+			"mediaForm", MediaForm.class);
+
+		long folderId = Optional.ofNullable(
+			mediaForm.getFolderId()
+		).orElse(
+			0L
+		);
+
+		BinaryFile binaryFile = multipartBody.getBinaryFile("file");
+
+		return _dlAppService.addFileEntry(
+			ddmFormInstance.getGroupId(), folderId, binaryFile.getFileName(),
+			binaryFile.getContentType(), mediaForm.getTitle(),
+			mediaForm.getDescription(), null, binaryFile.getInputStream(),
+			binaryFile.getSize(), new ServiceContext());
+	}
+
+	private Long _getFileEntryId(DDMFormFieldValue ddmFormFieldValue)
+		throws JSONException {
+
+		Value value = ddmFormFieldValue.getValue();
+
+		String valueString = value.getString(
+			LocaleThreadLocal.getDefaultLocale());
```

Comment:
this is strange as I already commented once... can't it be the locale in the API context? because it should be the requested language of the user or the default if he hasn't set one.